pkg/pushsync/pushsync.go

```diff
@@ -231,40 +231,41 @@ func (ps *PushSync) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream)
 
 	// forwarding replication
 	storedChunk := false
-	if ps.warmedUp() && ps.topologyDriver.IsWithinDepth(chunkAddress) {
-
-		verifiedChunk, err := ps.validStamp(chunk, ch.Stamp)
-		if err != nil {
-			ps.metrics.InvalidStampErrors.Inc()
-			ps.logger.Warningf("pushsync: forwarder, invalid stamp for chunk %s", chunkAddress.String())
-		} else {
-			chunk = verifiedChunk
-			_, err = ps.storer.Put(ctx, storage.ModePutSync, chunk)
-			if err != nil {
-				ps.logger.Warningf("pushsync: within depth peer's attempt to store chunk failed: %v", err)
-			} else {
-				storedChunk = true
+	defer func(stored *bool) {
+		if !*stored {
+			if ps.warmedUp() && ps.topologyDriver.IsWithinDepth(chunkAddress) {
+				verifiedChunk, err := ps.validStamp(chunk, ch.Stamp)
+				if err != nil {
+					ps.metrics.InvalidStampErrors.Inc()
+					ps.logger.Warningf("pushsync: forwarder, invalid stamp for chunk %s", chunkAddress.String())
+				} else {
+					chunk = verifiedChunk
```

Comment:
this was intended to be verified before forwarding with NN. with putting this in a defer, you're intervening in the sequence, and changing the protocol behavior. I think that having this code section in a defer is suboptimal. Having it somehow done at the end of the handler after all network operations have been finalized might be better.

Reply:
so far this was verified in sequence, however, with the only consequence that the error is logged. so this seems only to defer the storing without much impact for behaviour?